usb: fix TT bugs and class-specific functional descriptor discovery

adamgreloch opened the pull request:
JIRA: RTOS-1121

<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes shortly -->

Various bug fixes that came up when probing the CYW4373E dongle composed of a TT hub and 3 internal devices, some of which contained class-specific functional descriptors that were incorrectly parsed.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

These changes make it possible to run the CYW4373E driver (https://github.com/phoenix-rtos/phoenix-rtos-devices/pull/596).

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (refactoring, style fixes, git/CI config, submodule management, no code logic changes)

<!--- In case of breaking change - please advice here what needs to be done in dependent projects. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [X] Tested by hand on: nilee

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

adamgreloch committed:
usb/hub: fix deadlock on successful TT enumeration

JIRA: RTOS-1121

adamgreloch committed:
usb/dev: handle class-specific functional descriptors

JIRA: RTOS-1121

adamgreloch committed:
usb/hub: make TT workaround opt-in

When tested on imxrt1170-evkb with Laird/Sterling LWB5+ plugged in
(internal TT hub + two internal devices), the workaround for the lack of
interrupt pipe notifications was not neeeded. It seems that ia32 is the
only platform suffering from this issue.

JIRA: RTOS-1121

adamgreloch committed:
usb/dev: ignore failure to bind the driver

Returning failure on bind causes the discovery of subsequent devices on
a given hub to fail.  This is still yet to be fixed with proper orphans
management.

JIRA: RTOS-1121

adamgreloch commented on usb/hub.c:
```diff
@@ -426,8 +436,21 @@ static void hub_thread(void *args)
+			 * hub_ttStatus may lead to usb_devEnumerate call which may call hub_conf that
+			 * tries to obtain this lock. FIXME? It may be better to separate the enumeration
+			 * from the connect status polling logic, but as this polling loop is a
+			 * part of L2 TT workaround, the problem may be easier to solve itself in the root
```
@gemini-code-assist there is no deadlock. FIXME relates to suboptimal locking pattern (lock is unlocked then eventually locked inside in the recursion, unlocked and locked again on way back)

adamgreloch replied:
@gemini-code-assist what? `hub_ttStatus` does not hold a lock. There is not a single function call aside from `LIST_ADD` with `hub_common.lock` held.

adamgreloch commented on usb/hub.c:
```diff
@@ -495,14 +518,16 @@ int hub_conf(usb_dev_t *hub)
+		 * FIXME Interrupt pipe notifications from tier 2 TTs never come on ia32,
+		 * which is either a quirk in currently tested hw or an issue in the hcd driver.
+		 * As a workaround, try to receive interrupt requests AND actively poll the status.
+		 */
```
@gemini-code-assist this is literally the meaning of the FIXME comment...

adamgreloch commented on usb/dev.c:
```diff
@@ -346,7 +348,7 @@ static int usb_getConfiguration(usb_dev_t *dev)
 
 	for (size_t i = 0; i < dev->nifs; i++) {
-		if ((dev->ifs[i].desc == NULL) || (dev->ifs[i].eps == NULL)) {
+		if ((dev->ifs[i].desc == NULL) || ((dev->ifs[i].eps == NULL) && (dev->ifs[i].func == NULL))) {
```
It's not required for `eps`/`func` to be not NULL if `desc` is not NULL

julianuziemblo commented on usb/hub.c:
```diff
@@ -399,6 +407,7 @@ static uint32_t hub_getStatus(usb_dev_t *hub)
 }
 
 
+#if USB_HUB_TT_WORKAROUND
```
maybe also transfer this function higher in the file, so it can be under single `#if USB_HUB_TT_WORKAROUND` with `hub_isTT`, `hub_ttAdd` and `hub_ttRemove`?

julianuziemblo reviewed (commented):
Tested on `armv7m7-imxrt117x-evk`, DTR, works fine

one small nit/question, otherwise LGTM

ziemleszcz commented on usb/dev.c:
```diff
@@ -327,17 +327,19 @@ static int usb_getConfiguration(usb_dev_t *dev)
+					/* TODO: handle alternate setting maybe */
+				}
+				else {
+					dev->ifs[lastIfNum].func = (usb_generic_desc_t *)ptr;
```
What prevents this from accessing `dev->ifs[-1]`?

ziemleszcz commented on usb/hub.c:
```diff
@@ -426,8 +437,21 @@ static void hub_thread(void *args)
 			condWait(hub_common.cond, hub_common.lock, HUB_TT_POLL_DELAY_MS);
+
+			/*
+			 * hub_ttStatus may lead to usb_devEnumerate call which may call hub_conf that
```
Can you show the problematic path that leads to the mutex lock? I can see another path leading to `hub_ttRemove()`.
It appears that this lock protects both `hub_common.events` and `hub_common.tts` lists. Would introduction of separate locks for these lists fix this issue?

adamgreloch replied:
`hub_ttStatus -> hub_portstatus -> hub_connectstatus -> hub_devConnected -> usb_devEnumerate -> hub_conf -> hub_ttAdd`

The problem comes from the fact that `hub_portstatus` can be called from `hub_ttStatus` or from normal device discovery below, where it is assumed not to have the lock.

Actually, now I see that this workaround is still flawed (but to the other side), as `hub_ttStatus` now accesses `hub_common.tts` unsafely, so the mutex lock/unlock should be moved to `hub_ttStatus` for accessing the item from `hub_common.tts`.

A separate lock could help here a bit, but there is still a recursion with mutex over the same resource when calling `hub_ttStatus`. We would need a reentrant lock. Or deeply rework this architecture, but since this is just a workaround for x86 interrupt pipe problems, it may be easier to solve the root cause than that.

adamgreloch replied:
Moved the unlock/lock inside the `hub_ttStatus` and extended the comment

ziemleszcz commented:
Please combine two first commits.

ziemleszcz commented on usb/hub.c:
```diff
@@ -426,17 +449,12 @@ static void hub_thread(void *args)
-			 * part of L2 TT workaround, the problem may be easier to solve itself in the root
-			 */
-			mutexUnlock(hub_common.lock);
-			hub_ttStatus();
```
Another approach that avoids the deadlock is to iterate over all hubs and inject an artificial event for each, rather than invoking `hub_portstatus()`.

adamgreloch committed:
usb/hub: drop TT workaround

Since the TT interrupt problem concerns only the ia32 where USB is
experimentally supported anyways and is not reproducible on any other
targets, the source of the problem should be fixed on ia32 EHCI end, not
patched here.

JIRA: RTOS-1121

julianuziemblo commented:
I'd add additional description to [33214ba](https://github.com/phoenix-rtos/phoenix-rtos-usb/pull/38/commits/33214baab737661dd6068671209b57b0342f9acf) or to previous commit that it's left there for potential use in the future if it becomes necessary

adamgreloch commented:
> I'd add additional description to [33214ba](https://github.com/phoenix-rtos/phoenix-rtos-usb/pull/38/commits/33214baab737661dd6068671209b57b0342f9acf) or to previous commit that it's left there for potential use in the future if it becomes necessary

Git history already directly implies that
